Sprint 9 solution http api

tropuschka opened the pull request:
Здравствуйте!
Я помню, что задание нужно присылать законченное, но у меня упорно не складывается понимание, как сделать обработчики иначе и как состыковать их с тестами :(

tropuschka committed:
Добавить класс HttpTaskServer

tropuschka committed:
Создать сервер в классе HttpTaskServer

tropuschka committed:
Добавить BaseHttpHandler

tropuschka committed:
Добавить метода sendText в BaseHttpHandler

tropuschka committed:
Добавить метод sendNotFound в BaseHttpHandler

tropuschka committed:
Добавить метод sendHasInteractions в BaseHttpHandler

tropuschka committed:
Добавить обработчик для тасков (основа)

tropuschka committed:
Добавить обработчик для тасков (основа)

tropuschka committed:
Скорректировать код ответа в BaseHttpHandler.
Дополнить шаблон в TaskHttpHandler

tropuschka committed:
Добавить обработку неизвестных методов

tropuschka committed:
Создать контекст для разных случаев

tropuschka committed:
Создать обработчики (шаблоны) и обновить контекст

tropuschka committed:
Добавить библиотеку Gson

tropuschka committed:
Добавить создание менеджера и его привязку к обработчикам

tropuschka committed:
Добавить проверку строк на конвертируемость в число

tropuschka committed:
Обновить шаблон обработки запроса

tropuschka committed:
Добавить форматтер времени для обработчиков

tropuschka committed:
Добавить обработку метода POST

tropuschka committed:
Вынести создание тасков в отдельные методы в BaseHttpHandler

tropuschka committed:
Добавить обработку ошибок для метода POST

tropuschka committed:
Добавить проверку ошибок для метода DELETE

tropuschka committed:
Добавить обработку метода GET

tropuschka committed:
Переписала кучу всего, потому что оно не работало :)

tropuschka committed:
Переписать создание тасков

tropuschka committed:
Удалить лишние методы.
Почему-то менеджер не живет долго

tropuschka committed:
Добавить обработку для эпиков

tropuschka committed:
Добавить обработку для сабтасков

tropuschka committed:
Добавить обработку для истории. ОНа почему-то всегда пустая. Вероятно, потому что таски удаляются после первого же вызова

tropuschka committed:
Добавить обработку для тасков по приоритетности

tropuschka committed:
Пофиксить произвольное удаление тасков

tropuschka committed:
Удалить лишние экспорты

tropuschka committed:
Добавить метода для запуска сервера

tropuschka committed:
Не могу понять, как сделать тесты. Мне кажется, я что-то не так сделала в обработчиках, но придумать иначе не получается :(

tropuschka committed:
Проверки гита

tropuschka committed:
Проверки гита

tropuschka committed:
Проверки гита

tropuschka committed:
Проверки гита

tropuschka committed:
Проверки гита

Ivanov-s-tmn commented on src/httphandlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,183 @@
+                String param = item.substring(0, equalIndex);
+                String value = item.substring(equalIndex + 1);
+                switch (param) {
+                    case "name":
+                        taskName = value;
+                        break;
+                    case "description":
+                        taskDescription = value;
+                        break;
+                    case "start-date":
+                        dateString = value;
+                        break;
+                    case "end-date":
+                        endDateString = value;
+                        break;
+                    case "epic-id":
+                        if (isNumber(value)) epicId = Integer.parseInt(value);
+                        break;
+                }
+            }
+        }
+        if (dateString != null && endDateString != null) {
+            LocalDateTime taskStartDate = LocalDateTime.parse(dateString, formatter);
+            LocalDateTime taskEndDate = LocalDateTime.parse(endDateString, formatter);
+            Duration taskDuration = Duration.between(taskStartDate, taskEndDate);
+            durationString = taskDuration.toString();
+        }
+
+        switch (type) {
+            case "epic":
+                return new Epic(taskId, taskName, taskDescription);
+            case "subtask":
+                if (dateString != null)
+                    return new Subtask(taskId, taskName, taskDescription, epicId, dateString, durationString);
+                else return new Subtask(taskId, taskName, taskDescription, epicId);
+            default:
+                if (dateString != null) return new Task(taskId, taskName, taskDescription, dateString, durationString);
+                else return new Task(taskId, taskName, taskDescription);
+        }
+    }
```
Вообще, нужно воспользоваться gson для преобразования объекта в json и обратно. В http мы чаще всего используем формат json в теле запроса/ответа

Ivanov-s-tmn commented on src/httphandlers/EpicHttpHandler.java:
```diff
@@ -0,0 +1,82 @@
+                    int taskId = Integer.parseInt(requestArray[2]);
+                    task = (Epic) newTask(parameterString, "epic", taskId);
+                    Epic oldTask = manager.findEpicById(taskId);
+                    if (oldTask == null) sendNotFound(exchange, "Not Found");
+                    manager.updateEpic(task);
+                    if (Objects.equals(oldTask, manager.findEpicById(taskId)) || manager.findEpicById(taskId) == null) {
+                        sendHasInteractions(exchange, "Not Acceptable");
+                    }
+                    sendText(exchange, "Task \"" + task.getName() + "\" updated");
+                } else sendNotFound(exchange, "Not Found");
+                break;
+            case "GET":
+                if (requestArray[1].equals("epic") && requestArray.length == 2) {
+                    ArrayList<Epic> allTasksArray = manager.getAllEpic();
+                    if (allTasksArray.isEmpty()) sendNotFound(exchange, "Not Found");
+                    String response = allTasksArray.stream().map(Epic::toString).collect(Collectors.joining("\n"));
+                    sendText(exchange, response);
+                } else if (requestArray[1].equals("epic") && requestArray.length == 3 && isNumber(requestArray[2])) {
+                    int taskId = Integer.parseInt(requestArray[2]);
+                    task = manager.findEpicById(taskId);
+                    if (task == null) sendNotFound(exchange, "Not found");
+                    String response = task.toString();
+                    sendText(exchange, response);
+                } else sendNotFound(exchange, "Not Found");
+                break;
+            case "DELETE":
+                if (requestArray[1].equals("epic") && requestArray.length == 2) {
+                    manager.deleteAllEpics();
+                    sendText(exchange, "All tasks deleted");
+                } else if (requestArray[1].equals("epic") && requestArray.length == 3 && isNumber(requestArray[2])) {
+                    int taskId = Integer.parseInt(requestArray[2]);
+                    if (manager.findEpicById(taskId) == null) sendNotFound(exchange, "Not Found");
+                    String taskName = manager.findEpicById(taskId).getName();
+                    manager.deleteEpic(taskId);
+                    sendText(exchange, "Task \"" + taskName + "\" deleted");
+                } else sendNotFound(exchange, "Not Found");
+                break;
+            default:
+                sendNotFound(exchange, "Not Found");
+        }
```
Как пример реализации части метода обработки POST и GET запроса хендлера для эпиков для большего понимания:
```java
final String path = exchange.getRequestURI().getPath();
final Integer idFromRequest = getIdFromPath(path);
switch (exchange.getRequestMethod()) {
       case "GET": {
           if (idFromRequest == null) {
                    final List<Epic> tasks = taskManager.getEpics();
                    final String response = gson.toJson(tasks); //Преобразование в json для ответа
                    System.out.println("Получили все эпики");
                    sendText(exchange, response);
                    return;
            }
            final Epic epic = taskManager.getEpic(idFromRequest);
            if (epic != null) {
                if (path.contains("subtasks")) {
                     final List<Subtask> epicSubtasks = taskManager.getEpicSubtasks(idFromRequest);
                     final String response = gson.toJson(epicSubtasks);
                     System.out.println("Получили подзадачи эпика с id=" + idFromRequest);
                     sendText(exchange, response);
                } else {
                    final String response = gson.toJson(epic);
                    System.out.println("Получили эпик с id=" + idFromRequest);
                     sendText(exchange, response);
                }
                } else {
                    System.out.println("Эпика с id=" + idFromRequest + " не найдено");
                    sendNotFound(exchange);
                }
                break;
            }
        case "DELETE": {
                ...
            }
        case "POST": {
              String json = readText(exchange);
              final Epic epic = gson.fromJson(json, Epic.class); //преобразования из json с тела запроса
              final Integer id = epic.getId();
              if (id > 0) {
                   taskManager.updateEpic(epic);
                   System.out.println("Обновили эпик с id=" + id);
                   exchange.sendResponseHeaders(200, 0);
              } else {
                   try {
                        int idAdded = taskManager.addNewEpic(epic);
                        System.out.println("Создали эпик с id=" + idAdded);
                        final String response = gson.toJson(epic);
                        sendText(exchange, response);
                    } catch (TaskValidationException e) {
                        System.out.println("Задача пересекается с существующей");
                        sendHasInteractions(exchange);
                    }
                }
                break;
            }
       default: {
               ...
       }
}
```

Названия методов таск менеджера могут отличаться, я использую это как шаблон

Ivanov-s-tmn commented on java-kanban.iml:
Файл .iml локальный и его лучше добавить в gitignore

tropuschka committed:
Добавить класс getIdFromPath

tropuschka committed:
Обновить метод sendHasInteractions

tropuschka committed:
Удалить лишние методы

tropuschka committed:
Добавить исключение

tropuschka committed:
Добавить метод для преобразования мапы в таск

tropuschka committed:
Заменить метод для преобразования мапы в таск на конструктор

tropuschka committed:
Добавить метод для считывания таска из url

tropuschka committed:
Добавить обработку методов

tropuschka committed:
Добавить папку .github

tropuschka committed:
Обновить коды ответов

tropuschka committed:
Обновить обработчики для эпиков и сопутствующие методы

tropuschka committed:
Обновить обработчики для подзадач и сопутствующие методы

tropuschka committed:
Обновить обработчики для истории и списка по времени исполнения

tropuschka committed:
Обернуть обработчики в try-catch

tropuschka committed:
Обновить конструктор

tropuschka committed:
Добавить адаптер для обработки LocalDateTime

tropuschka committed:
Добавить тест

tropuschka committed:
Добавить адаптеры

tropuschka committed:
Заменить квери на джейсон

tropuschka commented:
Здравствуйте!
При запуске теста (он пока один) не сходится код ответа. Я попробовала поймать ошибку через дебаггер в разных классах, но не смогла :(
Подскажите, пожалуйста, где поискать проблему?

tropuschka committed:
Вернуть папку гитигнор

tropuschka committed:
Добавить тесты создания разных задач. Пофиксить кучу багов

tropuschka committed:
Добавить тесты обновления разных задач

tropuschka committed:
Добавить тесты получения разных задач

tropuschka committed:
gitignore

tropuschka committed:
Исправить получение айди из строки, добавить отклик на случай ненахождения таска

tropuschka committed:
Добавить тесты получения всех заданий

tropuschka committed:
Добавить тесты создания пересекающихся задач.
Для эпика таких тестов не делала, так как эпикам нельзя задать время при создании, но проверка в хэндлере на всякий случай есть

tropuschka committed:
Поправить проверку пересечений

tropuschka committed:
Добавить тест удаления отдельного таска

tropuschka committed:
Добавить тест удаления всех задач

tropuschka committed:
Добавить тест поиска несуществующих задач

tropuschka committed:
Добавить тест получения истории

tropuschka committed:
Добавить тест получения заданий, отсортированных по приоритету

tropuschka committed:
Почистить неиспользуемые экспорты и методы. Поправить формат дат

Ivanov-s-tmn commented on src/httphandlers/EpicHttpHandler.java:
```diff
@@ -17,66 +22,81 @@ public EpicHttpHandler(TaskManager manager) {
-                    manager.updateEpic(task);
-                    if (Objects.equals(oldTask, manager.findEpicById(taskId)) || manager.findEpicById(taskId) == null) {
-                        sendHasInteractions(exchange, "Not Acceptable");
+            Gson gson = new GsonBuilder()
+                    .registerTypeAdapter(LocalDateTime.class, new LocalDateTypeAdapter())
+                    .registerTypeAdapter(Duration.class, new DurationTypeAdapter())
+                    .registerTypeAdapter(DateTimeFormatter.class, new FormatterTypeAdapter())
+                    .create();
```
Можно инициализировать в базовом хендлере, чтобы не дублировать в каждой реализации

Ivanov-s-tmn commented on src/httphandlers/EpicHttpHandler.java:
```diff
@@ -17,66 +22,81 @@ public EpicHttpHandler(TaskManager manager) {
+                        return;
+                    }
+
+                    final Epic task = manager.findEpicById(requestedId);
```
Лишние вызовы методов менеджера не нужны. Это эндпоинт удаления, а вызывается дополнительно метод получения по айди, в нем есть лишняя логика, в нашем случае, это добавление в историю, что в данном случае не нужно. В каждом эндпоинте нужно вызывать только соответствующий ему метод менеджера. В остальных хендлерах аналогично

Ivanov-s-tmn commented on src/httphandlers/EpicHttpHandler.java:
```diff
@@ -17,66 +22,81 @@ public EpicHttpHandler(TaskManager manager) {
+                        sendSuccess(exchange);
+                    } else {
+                        int newId = manager.createEpic(task).getId();
+                        if (manager.findEpicById(newId) != null) {
```
Тут тоже лишний вызов. Аналогично в хендлерах задач и подзадач

Ivanov-s-tmn commented on src/taskmodels/Task.java:
```diff
@@ -3,15 +3,16 @@
     private TaskStatus status;
-    protected LocalDateTime startTime;
-    protected Duration duration;
+    public LocalDateTime startTime;
+    public Duration duration;
```
Поля не должны быть публичными, нужно соблюдать инкапсуляцию

Ivanov-s-tmn commented on src/taskmodels/Task.java:
```diff
@@ -31,6 +32,20 @@ public Task(Integer id, String name, String details, String startTime, String du
         this.duration = Duration.ofMinutes(Integer.parseInt(duration));
     }
 
+    public Task(Map<String, String> map, Integer id) {
+        if (id == null) this.id = 0;
+        this.name = map.get("name");
+        this.details = map.get("details");
+        this.status = TaskStatus.valueOf(map.get("status"));
+        String startTimeString = map.get("start-time");
+        String endTimeString = map.get("end-time");
+        if (startTimeString != null && endTimeString != null) {
+            this.startTime = LocalDateTime.parse(startTimeString);
+            LocalDateTime newEndTime = LocalDateTime.parse(endTimeString);
+            this.duration = Duration.between(this.startTime, newEndTime);
+        }
+    }
+
```
Очень странный конструктор, для чего он?)

tropuschka replied:
Был нужен раньше, когда не было json'ов, забыла про него)

tropuschka committed:
Удалить лишние определения переменной gson

tropuschka committed:
Удалить лишние конструкторы задач

tropuschka committed:
Заприватить параметры задач. Ранее gson не хотел с ними работать, когда они были приватными (сейчас работает)

tropuschka committed:
Удалить лишние вызовы поиска задач

Ivanov-s-tmn commented on java-kanban.iml:
.iml файл генерируется локально ide, его лучше добавить в исключения gitignore

Ivanov-s-tmn commented on test/HttpTaskServerTest.java:
В тестах дублируется создание тестовых задач и клиента, это все можно вынести в BeforeEach метод и сами тесты станут более компактными. Как правило, для тестов достаточно определенного набора тестовых данных, а остальное можно уже создать внутри тестов, по необходимости
